packages/ensjs/src/functions/transferSubname.ts

```
@@ -16,12 +17,46 @@ type NameWrapperArgs = {
 
 type Args = BaseArgs | NameWrapperArgs
 
+const transferWrappedSubname = async (
+  {
+    contracts,
+    signer,
+    getWrapperData,
+  }: ENSArgs<'contracts' | 'signer' | 'getWrapperData' | 'getExpiry'>,
+  name: string,
+  { contract, owner, resolverAddress, ...wrapperArgs }: NameWrapperArgs,
+) => {
+  const nameWrapper = (await contracts!.getNameWrapper()!).connect(signer)
+  const data = await getWrapperData(name)
+  const senderAddress = await signer.getAddress()
+
+  if (senderAddress === data?.owner) {
+    const node = namehash(name)
+    const resolver = await makeResolver({ contracts }, name, resolverAddress)
+    return nameWrapper.populateTransaction.setRecord(node, owner, resolver, 0)
+  }
+
+  const labels = name.split('.')
+  const label = labels.shift() as string
+  const parentNodehash = namehash(labels.join('.'))
+  const expiry = expiryToBigNumber(wrapperArgs.expiry, 0)
+
+  return nameWrapper.populateTransaction.setSubnodeOwner(
```

Instead of assuming that if the senderAddress is the parent owner , could we not batch the call for the owner with getting the parent owner and make sure the parent name is owned by the sender. Then we can just throw an error if the caller is neither.

---

We should probably also check for fuses here and if PCC is already burnt and if so we know we can't call setSubnodeOwner/Record

---

transferSubname implies the sender is the parent owner, we haven't done a prereq check like that yet

---

> Instead of assuming that if the senderAddress is the parent owner , could we not batch the call for the owner with getting the parent owner and make sure the parent name is owned by the sender. Then we can just throw an error if the caller is neither.

Currently the context code exists on the front end. We've had discussions about moving it into ensjs before but we didn't get to it in time.